plasmapy/analysis/swept_langmuir/ion_saturation_current.py

```diff
@@ -0,0 +1,240 @@
+          ``(voltage[extras.fitted_indices], current[extras.fitted_indices])``).
+
+    Notes
+    -----
+    This routine works by:
+
+    1. Selecting the points to be used in the fit as determined by
+       ``voltage_bound`` or ``current_bound``.
+    2. Fitting the selected points with the :term:`fit-function`
+       specified by ``fit_type``.
+    3. Extracting the linear component of the fit and returning that as
+       the ion-saturation current.
+
+    This routine opts to return a function representing a linear
+    ion-saturation current, since, while ideal planar Langmuir probes
+    reach a steady-state ion-saturation current, real world Langmuir
+    probes "suffer" from expanding sheaths as the bias voltage
+    becomes increasingly negative.  This sheath expansion results the
+    ion-saturation current also increasing.
+    """
+    rtn_extras = ISatExtras(rsq=None, fitted_func=None, fitted_indices=None)._asdict()
+
+    _settings = {
+        "linear": {
+            "func": ffuncs.Linear,
+            "current_bound": 0.4,
+        },
+        "exp_plus_linear": {
+            "func": ffuncs.ExponentialPlusLinear,
+            "current_bound": 1.0,
+        },
+        "exp_plus_offset": {
+            "func": ffuncs.ExponentialPlusOffset,
+            "current_bound": 1.0,
+        },
+    }  # type: Dict[str, Dict[str, Any]]
+    try:
+        default_current_bound = _settings[fit_type]["current_bound"]
+        fit_func = _settings[fit_type]["func"]()
+        rtn_extras["fitted_func"] = fit_func
+    except KeyError:
+        raise ValueError(
+            f"Requested fit '{fit_type}' is not a valid option.  Valid options "
+            f"are {list(_settings.keys())}."
+        )
+
+    # check voltage and current arrays
+    voltage, current = check_sweep(voltage, current, strip_units=True)
+
+    # condition kwargs voltage_bound and current_bound
+    if voltage_bound is None and current_bound is None:
+        current_bound = default_current_bound
+    elif voltage_bound is not None and current_bound is not None:
+        raise ValueError(
+            "Both keywords 'current_bound' and 'voltage_bound' are specified, "
+            "use only one."
+        )
+
+    if current_bound is not None:
+        if not isinstance(current_bound, numbers.Real):
+            raise TypeError(
+                f"Keyword 'current_bound' is of type {type(current_bound)}, "
+                f"expected an int or float."
+            )
+
+        current_min = current.min()
+        current_bound = (1.0 - current_bound) * current_min
+        mask = np.where(current <= current_bound)[0]
+    else:  # voltage_bound is not None
+        if not isinstance(voltage_bound, numbers.Real):
+            raise TypeError(
+                f"Keyword 'voltage_bound' is of type {type(voltage_bound)}, "
+                f"expected an int or float."
+            )
+
+        mask = np.where(voltage <= voltage_bound)[0]
```

Comment:
I wondered if [masked arrays from NumPy](https://numpy.org/doc/stable/reference/maskedarray.html) would be useful here, but it doesn't seem like changing the implementation would improve maintainability much.

Reply:
I'll have to read more about masked arrays, but I don't think it's quite applicable here. Because, `mask` is being used to filter two independent arrays, `voltage` and `current`.  I'm also suspecting, without definitive proof, that my current approach is probably less resource heavy.

Reply:
Cool!  I'm perfectly content keeping it as is.